lib/graphql/subscriptions/anycable_subscriptions.rb

```diff
@@ -141,34 +144,57 @@ def write_subscription(query, events)
+            events.each do |event|
+              pipeline.zincrby(redis_key(FINGERPRINTS_PREFIX) + event.topic, 1, event.fingerprint)
+              pipeline.sadd(redis_key(SUBSCRIPTIONS_PREFIX) + event.fingerprint, [subscription_id])
+            end
+            next unless config.subscription_expiration_seconds
+            pipeline.expire(redis_key(CHANNEL_PREFIX) + subscription_id, config.subscription_expiration_seconds)
+            pipeline.expire(redis_key(SUBSCRIPTION_PREFIX) + subscription_id, config.subscription_expiration_seconds)
           end
-          next unless config.subscription_expiration_seconds
-          pipeline.expire(redis_key(CHANNEL_PREFIX) + subscription_id, config.subscription_expiration_seconds)
-          pipeline.expire(redis_key(SUBSCRIPTION_PREFIX) + subscription_id, config.subscription_expiration_seconds)
         end
       end
 
       # Return the query from "storage" (in redis)
       def read_subscription(subscription_id)
-        redis.mapped_hmget(
-          "#{redis_key(SUBSCRIPTION_PREFIX)}#{subscription_id}",
-          :query_string, :variables, :context, :operation_name
-        ).tap do |subscription|
-          next if subscription.values.all?(&:nil?) # Redis returns hash with all nils for missing key
-
-          subscription[:context] = @serializer.load(subscription[:context])
-          subscription[:variables] = JSON.parse(subscription[:variables])
-          subscription[:operation_name] = nil if subscription[:operation_name].strip == ""
+        with_redis do |redis|
+          redis.mapped_hmget(
+            "#{redis_key(SUBSCRIPTION_PREFIX)}#{subscription_id}",
+            :query_string, :variables, :context, :operation_name
+          ).tap do |subscription|
+            next if subscription.values.all?(&:nil?) # Redis returns hash with all nils for missing key
+
+            subscription[:context] = @serializer.load(subscription[:context])
+            subscription[:variables] = JSON.parse(subscription[:variables])
+            subscription[:operation_name] = nil if subscription[:operation_name].strip == ""
+          end
+        end
+      end
+
+      # The channel was closed, forget about it and its subscriptions
+      def delete_channel_subscriptions(channel)
+        raise(ArgumentError, "Please pass channel instance to #{__method__} in your #unsubscribed method") if channel.is_a?(String)
+
+        channel_id = read_subscription_id(channel)
+
+        # Missing in case disconnect happens before #execute
+        return unless channel_id
+
+        with_redis do |redis|
+          redis.smembers(redis_key(CHANNEL_PREFIX) + channel_id).each do |subscription_id|
+            delete_subscription(redis, subscription_id)
+          end
+          redis.del(redis_key(CHANNEL_PREFIX) + channel_id)
         end
       end
 
-      def delete_subscription(subscription_id)
+      private
+
+      def delete_subscription(redis, subscription_id)
```

Comment:
Are you sure that `delete_subscription` should be private? And with changed arguments.

It doesn't [seem](https://github.com/search?q=repo%3Armosolgo%2Fgraphql-ruby delete_subscription&type=code) to be used outside of parent `GraphQL::Subscriptions` class though, so should be fine… but I'm not sure :sweat_smile:

Reply:
And just curious fact that following method declaration will work (and will be backward compatible!)
```suggestion
      def delete_subscription(subscription_id, redis: with_redis { |conn| conn }))
```